library/pkcs7_write.c

```diff
@@ -0,0 +1,829 @@
+/* Copyright 2020 IBM Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "mbedtls/config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+#if defined(MBEDTLS_PKCS7_WRITE_C)
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "mbedtls/asn1write.h"  // for building pkcs7
+#include "mbedtls/base64.h"     // for pem 2 der functions
+#include "mbedtls/md.h"         //  generic interface
+#include "mbedtls/md_internal.h"
+#include "mbedtls/oid.h"
+#include "mbedtls/pk_internal.h"
+#include "mbedtls/pkcs7.h"
+#include "mbedtls/pkcs7_write.h"
+#include "mbedtls/platform.h"  //  mbedtls functions
+#include "mbedtls/x509_crt.h"
+#if defined(MBEDTLS_FS_IO)
+#include <sys/stat.h>
+#include <sys/types.h>
+#endif
+#if defined(MBEDTLS_PLATFORM_C)
+#include "mbedtls/platform.h"
+#include "mbedtls/platform_util.h"
+#else
+#include <stdio.h>
+#include <stdlib.h>
+#define mbedtls_free free
+#define mbedtls_calloc calloc
+#define mbedtls_printf printf
+#define mbedtls_snprintf snprintf
+#endif
+
+/* STRUCTURE OF PKCS7 AND CORRESPONDING FUNCTIONS THAT HANDLE THEM:
+ *PKCS7 {
+ *	CONSTRUCTED | SEQUENCE 										->pkcs7_set_pkcs7_oid
+ *		OID (Signed Data) 										->^
+ *		CONSTRUCTED | CONTEXT SPECIFIC 							->pkcs7_set_signed_data
+ *			CONSTRUCTED | SEQUENCE 								->pkcs7_set_version
+ *				INTEGER (version) 								->^
+ *				CONSTRUCTED |SET 								->pkcs7_set_algo_id
+ *					OID (hash Alg)								->^
+ *				CONSTRUCTED | SEQUENCE 							->pkcs7_set_signed_data_oid
+ *					OID (PKCS7 Data)							->^
+ *				CONSTRUCTED | CONTEXT SPECIFIC 					->pkcs7_set_signer_certs
+ *					entire certificate for each signer  		->^
+ *				CONSTRUCTED | SET 								->pkcs7_set_signers_data
+ *					CONSTRUCTED | SEQUENCE (for each signer) 	->pkcs7_set_each_signer_data
+ *						INTEGER (signedInfoVersion) 			->pkcs7_set_signer_cert_data
+ *						CONSTRUCTED | SEQUENCE 					->^
+ *							certificate issuer info 			->^
+ *							certificate serial					->^
+ *						CONSTRUCTED | SEQUENCE 					->pkcs7_set_algorithm_ids
+ *							OID (hash Alg)						->^
+ *						CONSTRUCTED | SEQUENCE 					->^
+ *							OID (Signature Alg (RSA)) 			->^
+ *						OCTET STRING (signature) 				->pkcs7_set_signature
+ * }
+ */
+
+//struct for containing information needed to generate the pkcs7
```

Comment:
Style: we tend to use `/*` comments. Not a big deal, as we're officially allowing C99 in the development branch, but I still find this more readable if the same style is used everywhere.